File: sdk/go/device.go
package conduyt

import (
	"context"
	"encoding/binary"
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/virgilvox/conduyt/sdk/go/transports"
)

// Device is a high-level CONDUYT device client.
type Device struct {
	transport transports.Transport
	timeout   time.Duration
	seq       byte
	mu        sync.Mutex
	pending   map[byte]chan *Packet
	hello     *HelloResp
	done      chan struct{}
}

// NewDevice creates a new Device with the given transport.
func NewDevice(t transports.Transport, timeout time.Duration) *Device {
	if timeout == 0 {
		timeout = 5 * time.Second
	}
	return &Device{
		transport: t,
		timeout:   timeout,
		pending:   make(map[byte]chan *Packet),
		done:      make(chan struct{}),
	}
}

// Connect performs the HELLO handshake.
func (d *Device) Connect(ctx context.Context) (*HelloResp, error) {
	if err := d.transport.Connect(); err != nil {
		return nil, fmt.Errorf("connect: %w", err)
	}

	go d.recvLoop()

	resp, err := d.sendCommand(ctx, CmdHello, nil)
	if err != nil {
		return nil, fmt.Errorf("hello: %w", err)
	}

	hello, err := ParseHelloResp(resp.Payload)
	if err != nil {
		return nil, fmt.Errorf("parse hello: %w", err)
	}
	d.hello = hello
	return hello, nil
}

// Close closes the device connection.
func (d *Device) Close() error {
	close(d.done)
	d.mu.Lock()
	for _, ch := range d.pending {
		close(ch)
	}
	d.pending = make(map[byte]chan *Packet)
	d.mu.Unlock()
	return d.transport.Close()
}

// Ping sends a PING and waits for PONG.
func (d *Device) Ping(ctx context.Context) error {
	_, err := d.sendCommand(ctx, CmdPing, nil)
	return err
}

// Reset sends a RESET command.
func (d *Device) Reset(ctx context.Context) error {
	_, err := d.sendCommand(ctx, CmdReset, nil)
	return err
}

// Pin returns a PinProxy for the given pin number.
func (d *Device) Pin(n byte) *PinProxy {
	return &PinProxy{device: d, pin: n}
}

// PinByName returns a PinProxy from a name like "A0", "A5", or "13".
// Analog names (A0-A15) auto-set the proxy to analog mode.
func (d *Device) PinByName(name string) (*PinProxy, error) {
	m := analogPinRe.FindStringSubmatch(name)
	if m != nil {
		n, _ := strconv.Atoi(m[1])
		return &PinProxy{device: d, pin: byte(n), isAnalog: true}, nil
	}
	n, err := strconv.Atoi(strings.TrimSpace(name))
	if err != nil {
		return nil, fmt.Errorf("invalid pin name: %q", name)
	}
	return &PinProxy{device: d, pin: byte(n)}, nil
}

// Capabilities returns the parsed HELLO_RESP, or nil if not connected.
func (d *Device) Capabilities() *HelloResp {
	return d.hello
}

// ModCmd sends a MOD_CMD and waits for ACK/MOD_RESP. Returns the response
// payload (empty for ACK, populated for MOD_RESP). Module commands that
// only ACK can ignore the byte slice; commands like Encoder.Read or
// DHT.Read need the bytes.
func (d *Device) ModCmd(ctx context.Context, payload []byte) ([]byte, error) {
	resp, err := d.sendCommand(ctx, CmdModCmd, payload)
	if err != nil {
		return nil, err
	}
	if resp == nil {
		return nil, nil
	}
	return resp.Payload, nil
}

// OTABegin starts an OTA update. sha256 must be exactly 32 bytes. The
// firmware allocates space + verifies the digest at finalize time.
func (d *Device) OTABegin(ctx context.Context, totalBytes uint32, sha256 []byte) error {
	if len(sha256) != 32 {
		return fmt.Errorf("OTABegin: sha256 must be 32 bytes, got %d", len(sha256))
	}
	payload := make([]byte, 36)
	binary.LittleEndian.PutUint32(payload[0:4], totalBytes)
	copy(payload[4:36], sha256)
	_, err := d.sendCommand(ctx, CmdOTABegin, payload)
	return err
}

// OTAChunk sends one OTA chunk. Offsets must be sequential — the firmware
// NAKs out-of-order chunks with OTA_INVALID.
func (d *Device) OTAChunk(ctx context.Context, offset uint32, data []byte) error {
	payload := make([]byte, 4+len(data))
	binary.LittleEndian.PutUint32(payload[0:4], offset)
	copy(payload[4:], data)
	_, err := d.sendCommand(ctx, CmdOTAChunk, payload)
	return err
}

// OTAFinalize finalizes the update. Firmware verifies the SHA256 and reboots
// into the new image. The next round-trip will fail until you reconnect.
func (d *Device) OTAFinalize(ctx context.Context) error {
	_, err := d.sendCommand(ctx, CmdOTAFinalize, nil)
	return err
}

// sendCommand sends a command and waits for the response.
func (d *Device) sendCommand(ctx context.Context, cmdType byte, payload []byte) (*Packet, error) {
	d.mu.Lock()
	seq := d.seq
	d.seq++
	ch := make(chan *Packet, 1)
	d.pending[seq] = ch
	d.mu.Unlock()

	pkt := MakePacket(cmdType, seq, payload)
	encoded := WireEncode(pkt)

	if err := d.transport.Send(encoded); err != nil {
		d.mu.Lock()
		delete(d.pending, seq)
		d.mu.Unlock()
		return nil, err
	}

	ctx2, cancel := context.WithTimeout(ctx, d.timeout)
	defer cancel()

	select {
	case resp, ok := <-ch:
		if !ok {
			return nil, fmt.Errorf("connection closed")
		}
		if resp.Type == EvtNAK {
			code := byte(0)
			if len(resp.Payload) > 0 {
				code = resp.Payload[0]
			}
			return nil, NewNAKError(code, seq)
		}
		return resp, nil
	case <-ctx2.Done():
		d.mu.Lock()
		delete(d.pending, seq)
		d.mu.Unlock()
		return nil, &TimeoutError{Seq: seq, TimeoutMs: int(d.timeout.Milliseconds())}
	}
}

func (d *Device) recvLoop() {
	for {
		select {
		case <-d.done:
			return
		default:
		}

		raw, err := d.transport.Recv()
		if err != nil {
			return
		}

		pkt, err := WireDecode(raw)
		if err != nil {
			continue
		}

		d.mu.Lock()
		ch, ok := d.pending[pkt.Seq]
		if ok {
			delete(d.pending, pkt.Seq)
		}
		d.mu.Unlock()

		if ok {
			ch <- pkt
		}
	}
}

File: sdk/rust/src/device.rs
//! CONDUYT Device — async-free device client

extern crate alloc;
use alloc::vec::Vec;

use crate::constants::*;
use crate::wire::{Packet, wire_encode, wire_decode, make_packet};
use crate::transports::Transport;

/// Error from device operations.
#[derive(Debug)]
pub enum DeviceError<E: core::fmt::Debug> {
    Transport(E),
    Nak { code: u8, name: &'static str, seq: u8 },
    Timeout { seq: u8 },
    WireError(crate::wire::WireError),
    NotConnected,
}

impl<E: core::fmt::Debug> core::fmt::Display for DeviceError<E> {
    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
        match self {
            Self::Transport(e) => write!(f, "transport error: {:?}", e),
            Self::Nak { code, name, seq } => write!(f, "NAK: {} (0x{:02x}, seq={})", name, code, seq),
            Self::Timeout { seq } => write!(f, "timeout: seq={}", seq),
            Self::WireError(e) => write!(f, "wire: {}", e),
            Self::NotConnected => write!(f, "not connected"),
        }
    }
}

/// Synchronous CONDUYT device client.
pub struct Device<T: Transport> {
    transport: T,
    seq: u8,
}

impl<T: Transport> Device<T> {
    /// Create a new device client.
    pub fn new(transport: T) -> Self {
        Self { transport, seq: 0 }
    }

    /// Borrow the underlying transport. Useful for tests that need to inspect
    /// the wire bytes a `MockTransport` captured.
    pub fn transport(&self) -> &T { &self.transport }

    /// Connect and perform HELLO handshake.
    pub fn connect(&mut self) -> Result<Vec<u8>, DeviceError<T::Error>> {
        self.transport.connect().map_err(DeviceError::Transport)?;
        let resp = self.send_command(CMD_HELLO, &[])?;
        Ok(resp.payload)
    }

    /// Send a PING.
    pub fn ping(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.send_command(CMD_PING, &[])?;
        Ok(())
    }

    /// Send a RESET.
    pub fn reset(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.send_command(CMD_RESET, &[])?;
        Ok(())
    }

    /// Close the connection.
    pub fn close(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.transport.close().map_err(DeviceError::Transport)
    }

    /// Set pin mode.
    pub fn pin_mode(&mut self, pin: u8, mode: u8) -> Result<(), DeviceError<T::Error>> {
        self.send_command(CMD_PIN_MODE, &[pin, mode])?;
        Ok(())
    }

    /// Write to a pin.
    pub fn pin_write(&mut self, pin: u8, value: u8) -> Result<(), DeviceError<T::Error>> {
        self.send_command(CMD_PIN_WRITE, &[pin, value])?;
        Ok(())
    }

    /// Read a pin value using the mode stored on the firmware
    /// (set via `pin_mode`). Defaults to digital if no mode was set.
    pub fn pin_read(&mut self, pin: u8) -> Result<u16, DeviceError<T::Error>> {
        let resp = self.send_command(CMD_PIN_READ, &[pin])?;
        if resp.payload.len() >= 3 {
            Ok(u16::from_le_bytes([resp.payload[1], resp.payload[2]]))
        } else {
            Ok(0)
        }
    }

    /// Read analog value (0-1023 for 10-bit ADC).
    /// Sends the ANALOG mode byte so the firmware calls analogRead().
    pub fn analog_read(&mut self, pin: u8) -> Result<u16, DeviceError<T::Error>> {
        let resp = self.send_command(CMD_PIN_READ, &[pin, PIN_MODE_ANALOG])?;
        if resp.payload.len() >= 3 {
            Ok(u16::from_le_bytes([resp.payload[1], resp.payload[2]]))
        } else {
            Ok(0)
        }
    }

    /// Read digital value (0 or 1).
    /// Sends the INPUT mode byte so the firmware calls digitalRead().
    pub fn digital_read(&mut self, pin: u8) -> Result<u16, DeviceError<T::Error>> {
        let resp = self.send_command(CMD_PIN_READ, &[pin, PIN_MODE_INPUT])?;
        if resp.payload.len() >= 3 {
            Ok(u16::from_le_bytes([resp.payload[1], resp.payload[2]]))
        } else {
            Ok(0)
        }
    }

    /// Send a module command.
    /// Begin an OTA update. `sha256` must be exactly 32 bytes.
    pub fn ota_begin(&mut self, total_bytes: u32, sha256: &[u8]) -> Result<(), DeviceError<T::Error>> {
        if sha256.len() != 32 {
            // Reuse WireError as a generic argument-shape error.
            return Err(DeviceError::WireError(crate::wire::WireError::IncompletePkt));
        }
        let mut payload = [0u8; 36];
        payload[0..4].copy_from_slice(&total_bytes.to_le_bytes());
        payload[4..36].copy_from_slice(sha256);
        self.send_command(CMD_OTA_BEGIN, &payload)?;
        Ok(())
    }

    /// Send one OTA chunk. Offsets must be sequential — the firmware NAKs
    /// out-of-order chunks with OTA_INVALID.
    pub fn ota_chunk(&mut self, offset: u32, data: &[u8]) -> Result<(), DeviceError<T::Error>> {
        let mut payload = Vec::with_capacity(4 + data.len());
        payload.extend_from_slice(&offset.to_le_bytes());
        payload.extend_from_slice(data);
        self.send_command(CMD_OTA_CHUNK, &payload)?;
        Ok(())
    }

    /// Finalize the OTA update. Firmware verifies the SHA256 + reboots.
    pub fn ota_finalize(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.send_command(CMD_OTA_FINALIZE, &[])?;
        Ok(())
    }

    pub fn mod_cmd(&mut self, payload: &[u8]) -> Result<Vec<u8>, DeviceError<T::Error>> {
        let resp = self.send_command(CMD_MOD_CMD, payload)?;
        Ok(resp.payload)
    }

    /// Send a command and wait for response.
    fn send_command(&mut self, cmd_type: u8, payload: &[u8]) -> Result<Packet, DeviceError<T::Error>> {
        if !self.transport.connected() {
            return Err(DeviceError::NotConnected);
        }

        let seq = self.seq;
        self.seq = self.seq.wrapping_add(1);

        let pkt = make_packet(cmd_type, seq, payload);
        let encoded = wire_encode(&pkt);

        self.transport.send(&encoded).map_err(DeviceError::Transport)?;

        let raw = self.transport.recv().map_err(DeviceError::Transport)?;
        let resp = wire_decode(&raw).map_err(DeviceError::WireError)?;

        if resp.pkt_type == EVT_NAK {
            let code = if resp.payload.is_empty() { 0 } else { resp.payload[0] };
            return Err(DeviceError::Nak {
                code,
                name: err_name(code),
                seq,
            });
        }

        Ok(resp)
    }
}
